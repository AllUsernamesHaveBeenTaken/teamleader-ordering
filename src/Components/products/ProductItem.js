import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { OrderContext } from '../../Contexts';

import { Icon } from '../../Elements';

const ProductItem = ({ item, orderId, success }) => (
  <OrderContext.Consumer>
    {context => (
      <TableRow>
        <td>{item['product-id']}</td>
        <td>
          {!success && (
            <button
              type="submit"
              disabled={success}
              onClick={() => context.substractItem(orderId, item['product-id'])}
            >
              -
            </button>
          )}

          {item.quantity}

          {!success && (
            <button
              type="submit"
              disabled={success}
              onClick={() => context.addItem(orderId, item['product-id'])}
            >
              +
            </button>
          )}
        </td>
        <td>{item['unit-price']}</td>
        <td>{item.total}</td>
        {!success && (
          <td>
            <IconWrapper onClick={() => context.deleteItem(orderId, item['product-id'])}>
              <Icon name="trash" color="black" />
            </IconWrapper>
          </td>
        )}
      </TableRow>
    )}
  </OrderContext.Consumer>
);

export default ProductItem;

ProductItem.propTypes = {
  item: PropTypes.shape({
    'product-id': PropTypes.string.isRequired,
    quantity: PropTypes.string.isRequired,
    'unit-price': PropTypes.string.isRequired,
    total: PropTypes.string.isRequired,
  }).isRequired,
  orderId: PropTypes.string.isRequired,
  success: PropTypes.bool,
};

ProductItem.defaultProps = {
  success: false,
};

const TableRow = styled.tr`
  border-bottom: 1px solid black;
`;

const IconWrapper = styled.div`
  cursor: pointer;
`;
