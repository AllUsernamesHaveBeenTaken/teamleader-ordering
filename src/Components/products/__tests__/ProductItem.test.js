import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import ProductItem from '../ProductItem';

afterEach(() => {
  cleanup();
  console.error.mockClear();
});

console.error = jest.fn();

const substractItem = jest.fn();
const addItem = jest.fn();
const deleteItem = jest.fn();

const item = {
  'product-id': 'B102',
  quantity: '10',
  total: '49.90',
  'unit-price': '4.99',
};

const orderId = '1';

// ProductItem's props = item, orderId, success, substractItem, addItem, deleteItem
test('<ProductItem /> where success is true', () => {
  const success = true;
  const { queryByTestId, getByTestId } = render(
    <ProductItem
      item={item}
      orderId={orderId}
      success={success}
      substractItem={substractItem}
      addItem={addItem}
      deleteItem={deleteItem}
    />,
  );

  expect(getByTestId('item-row')).toBeTruthy();

  expect(queryByTestId('substract-item')).toBeFalsy();
  expect(substractItem).toHaveBeenCalledTimes(0);

  expect(queryByTestId('add-item')).toBeFalsy();
  expect(addItem).toHaveBeenCalledTimes(0);

  expect(queryByTestId('delete-item')).toBeFalsy();
  expect(deleteItem).toHaveBeenCalledTimes(0);
});

test('<ProductItem /> where success is false', () => {
  const success = false;
  const { getByTestId } = render(
    <ProductItem
      item={item}
      orderId={orderId}
      success={success}
      substractItem={substractItem}
      addItem={addItem}
      deleteItem={deleteItem}
    />,
  );

  expect(getByTestId('item-row')).toBeTruthy();

  fireEvent.click(getByTestId('substract-item'));
  expect(substractItem).toHaveBeenCalledTimes(1);
  expect(substractItem).toHaveBeenCalledWith(orderId, item['product-id']);

  fireEvent.click(getByTestId('add-item'));
  expect(addItem).toHaveBeenCalledTimes(1);
  expect(addItem).toHaveBeenCalledWith(orderId, item['product-id']);

  fireEvent.click(getByTestId('delete-item'));
  expect(deleteItem).toHaveBeenCalledTimes(1);
  expect(deleteItem).toHaveBeenCalledWith(orderId, item['product-id']);
});
