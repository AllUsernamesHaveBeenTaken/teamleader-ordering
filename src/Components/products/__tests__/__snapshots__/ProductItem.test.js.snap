// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> where success is false 1`] = `
<tr
  class="sc-bZQynM bGQAfj"
  data-testid="item-row"
>
  <td>
    B102
  </td>
  <td>
    <button
      type="submit"
      data-testid="substract-item"
    >
      -
    </button>
    10
    <button
      type="submit"
      data-testid="add-item"
    >
      +
    </button>
  </td>
  <td>
    4.99
  </td>
  <td>
    49.90
  </td>
  <td>
    <div
      class="sc-gzVnrw btWNie"
      data-testid="delete-item"
    >
      <svg
        width="24"
        height="24"
        viewBox="0 0 24 24"
      >
        <path
          fill="black"
          d="M9 19c0 .552-.448 1-1 1s-1-.448-1-1v-10c0-.552.448-1 1-1s1 .448 1 1v10zm4 0c0 .552-.448 1-1 1s-1-.448-1-1v-10c0-.552.448-1 1-1s1 .448 1 1v10zm4 0c0 .552-.448 1-1 1s-1-.448-1-1v-10c0-.552.448-1 1-1s1 .448 1 1v10zm5-17v2h-20v-2h5.711c.9 0 1.631-1.099 1.631-2h5.315c0 .901.73 2 1.631 2h5.712zm-3 4v16h-14v-16h-2v18h18v-18h-2z"
        />
      </svg>
    </div>
  </td>
</tr>
`;
